src/internal/operators/sequenceEqual.ts

```diff
@@ -62,113 +61,117 @@ import { lift } from '../util/lift';
+            sequenceEqualSubscriber?.unsubscribe();
+          }
+        );
+
+        return sequenceEqualSubscriber;
+      };
+
+      // Subscribe to each source.
+      source.subscribe(createSubscriber(aState, bState));
+      compareTo.subscribe(createSubscriber(bState, aState));
+    });
 }
 
-export class SequenceEqualOperator<T> implements Operator<T, boolean> {
-  constructor(private compareTo: Observable<T>,
-              private comparator?: (a: T, b: T) => boolean) {
-  }
-
-  call(subscriber: Subscriber<boolean>, source: any): any {
-    return source.subscribe(new SequenceEqualSubscriber(subscriber, this.compareTo, this.comparator));
-  }
+/**
+ * A simple structure for the data used to test each sequence
+ */
+interface SequenceState<T> {
+  /** A temporary store for arrived values before they are checked */
+  buffer: T[];
+  /** Whether or not the sequence source has completed. */
+  complete: boolean;
 }
 
 /**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
+ * Creates a simple structure that is used to represent
+ * data used to test each sequence.
  */
-export class SequenceEqualSubscriber<T, R> extends Subscriber<T> {
-  private _a: T[] = [];
-  private _b: T[] = [];
-  private _oneComplete = false;
-
-  constructor(destination: Observer<R>,
-              private compareTo: Observable<T>,
-              private comparator?: (a: T, b: T) => boolean) {
-    super(destination);
-    (this.destination as Subscription).add(compareTo.subscribe(new SequenceEqualCompareToSubscriber(destination, this)));
-  }
-
-  protected _next(value: T): void {
-    if (this._oneComplete && this._b.length === 0) {
-      this.emit(false);
-    } else {
-      this._a.push(value);
-      this.checkValues();
-    }
-  }
-
-  public _complete(): void {
-    if (this._oneComplete) {
-      this.emit(this._a.length === 0 && this._b.length === 0);
-    } else {
-      this._oneComplete = true;
-    }
-    this.unsubscribe();
-  }
-
-  checkValues() {
-    const { _a, _b, comparator } = this;
-    while (_a.length > 0 && _b.length > 0) {
-      let a = _a.shift()!;
-      let b = _b.shift()!;
-      let areEqual = false;
-      try {
-        areEqual = comparator ? comparator(a, b) : a === b;
-      } catch (e) {
-        this.destination.error(e);
-      }
-      if (!areEqual) {
-        this.emit(false);
-      }
-    }
-  }
-
-  emit(value: boolean) {
-    const { destination } = this;
-    destination.next(value);
-    destination.complete();
-  }
-
-  nextB(value: T) {
-    if (this._oneComplete && this._a.length === 0) {
-      this.emit(false);
-    } else {
-      this._b.push(value);
-      this.checkValues();
-    }
-  }
-
-  completeB() {
-    if (this._oneComplete) {
-      this.emit(this._a.length === 0 && this._b.length === 0);
-    } else {
-      this._oneComplete = true;
-    }
-  }
+function createState<T>(): SequenceState<T> {
+  return {
+    buffer: [] as T[],
```

Comment:
nitpick: Is this `as T[]` assertion really needed? I thought empty arrays were typed as `never[]` which ought to be assignable to any array type.